### PR TITLE
Solves issue 7153 

### PR DESCRIPTION
**Pull request by vruano**
involving  "zombie" likelihoods from past removed evidences being assigned to new appended evidences in AlleleLikelihoods.

Refactor and clean some code as well.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/models/AlleleFractionLikelihoods.java [hunk 1]
```diff
@@ -27,7 +27,7 @@
  * We need such functionality because MCMC evaluates the likelihood under proposed parameter changes.
  *
  * <p>
- * if indicator == ALT_MINOR:
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/models/AlleleFractionLikelihoods.java:30]: Don't know why I did this :P

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/AlleleLikelihoodMatrixMapper.java [hunk 2]
```diff
@@ -58,7 +55,7 @@ public double get(final int alleleIndex, final int evidenceIndex) {
             }
 
             @Override
-            public int indexOfAllele(final A allele) {
+            public int indexOfAllele(final Allele allele) {
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/AlleleLikelihoodMatrixMapper.java:58]: This change is totally legal (e.g. see Map.containsKey) and fixes supress-warning issues in other places. Problem is that it cascaded to all descendants and that. is the reason this PR touches so many files.

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java [hunk 3]
```diff
@@ -446,7 +445,7 @@ private void normalizeLikelihoodsPerEvidence(final double maximumBestAltLikeliho
      * @return never {@code null}, but with {@link BestAllele#allele allele} == {@code null}
      * if non-could be found.
      */
-    private BestAllele searchBestAllele(final int sampleIndex, final int evidenceIndex, final boolean canBeReference, Optional<double[]> priorities) {
+    private BestAllele searchBestAllele(final int sampleIndex, final int evidenceIndex, final boolean canBeReference, final double[] priorities) {
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java:448]: My IntelliJ complains about using Optional on a method argument... don't know why that is smelly. In any case a nullable value covers the same functionality nicely and is more efficient, I guess.

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java [hunk 4]
```diff
@@ -848,15 +842,9 @@ public void updateNonRefAlleleLikelihoods() {
...
      * Updates the likelihood of the NonRef allele (if present) based on the likelihoods of a set of non-symbolic
-     * <p>
-     *     This method does
-     * </p>
-     *
-     * @param allelesToConsider
      */
-    @SuppressWarnings("unchecked")  // for the cast (A) Allele.NON_REF_ALLELE below
     public void updateNonRefAlleleLikelihoods(final AlleleList<A> allelesToConsider) {
-        final int nonRefAlleleIndex = indexOfAllele((A) Allele.NON_REF_ALLELE);
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java:859]: This is the unchecked cast that is fixed.

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java [hunk 5]
```diff
@@ -1180,9 +1169,12 @@ private void removeEvidenceByIndex(final int sampleIndex, final int[] evidencesT
                 }
             }
         }
+        // set to NaN lks of the deleted positions in lk value arrays.
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java:1172]: I'm proactively getting rid of the zombie lks when evidence is removed. I set them to NaN.

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java [hunk 6]
```diff
@@ -1180,9 +1169,12 @@ private void removeEvidenceByIndex(final int sampleIndex, final int[] evidencesT
...
+        for (final double[] alleleValues : valuesBySampleIndex[sampleIndex]) {
+            Arrays.fill(alleleValues, newEvidenceCount, alleleValues.length, Double.NaN);
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java:1174]: actually the to index could be ```oldEvidenceCount``` instead. Will change.

File: src/test/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoodsUnitTest.java [hunk 7]
```diff
@@ -765,21 +787,24 @@ private void testSampleQueries(final SampleList samples, Map<String, List<GATKRe
...
-            {},
-            { 100 },
-            { 0 },
-            { 0, 0, 0 },
-            { 1, 0, 1 },
-            { 100, 10 , 100},
-            { 1000, 10, 100, 20, 23 }
-    };
+    @Test
+    public void testBug7153ZoombeeLkValues() {
```
Comment (vruano) [inline src/test/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoodsUnitTest.java:791]: Terrible spelling :S

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java [hunk 8]
```diff
@@ -776,17 +775,12 @@ public void addEvidence(final Map<String,List<EVIDENCE>> evidenceBySample, final
     }
 
     // Extends the likelihood arrays-matrices.
-    private void extendsLikelihoodArrays(final double initialLikelihood, final int sampleIndex, final int sampleEvidenceCount, final int newSampleEvidenceCount) {
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java:779]: This is the fixing change right below.

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/LikelihoodMatrix.java [hunk 9]
```diff
@@ -14,13 +14,13 @@
      * List of evidence in the matrix sorted by their index therein.
      * @return never {@code null}.
      */
-    public List<EVIDENCE> evidence();
+    List<EVIDENCE> evidence();
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/LikelihoodMatrix.java:17]: Do you mean to get rid of the `public`s here?

**Review by davidbenjamin — APPROVED** [review]
Looks good, just wondering about removing `public` in a few places.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/AlleleLikelihoodMatrixMapper.java [hunk 10]
```diff
@@ -10,23 +10,20 @@
 /**
  * Creates {@link org.broadinstitute.hellbender.utils.genotyper.LikelihoodMatrix} mappers to be used when working with a subset of the original alleles.
  */
-public class AlleleLikelihoodMatrixMapper<A extends Allele> {
+class AlleleLikelihoodMatrixMapper<A extends Allele> {
```
[the remark this answers is not in the document]
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/AlleleLikelihoodMatrixMapper.java:13]: Usually I follow Intellj suggestions in downgrading visitibility... if is not used yet is better to keep it hidden.

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/LikelihoodMatrix.java (hunk 9 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/utils/genotyper/LikelihoodMatrix.java:17]: in Interfaces public is implied.

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java (hunk 6 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java:1174]: I refactored the code a bit to make a bit clearer what is going on.

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/LikelihoodMatrix.java (hunk 9 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/genotyper/LikelihoodMatrix.java:17]: Ah, of course!  I forgot it was an interface.

**Comment by davidbenjamin** [comment]
@vruano Good to merge whenever you want!

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java [hunk 11]
```diff
@@ -63,12 +63,18 @@ private double getInformativeThreshold() {
      *     valuesBySampleIndex[s][a][r] == lnLk(R_r | A_a) where R_r comes from Sample s.
      * </p>
      */
-    protected final double[][][] valuesBySampleIndex;
+    private final double[][][] valuesBySampleIndex;
+
+    /**
+     * Keeps track of the maximum number of evidences and likelihood values that we cans
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java:69]: we cans -> can be stored.

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java [hunk 12]
```diff
@@ -776,17 +794,29 @@ public void addEvidence(final Map<String,List<EVIDENCE>> evidenceBySample, final
...
+    // Resizes the lk value holding arrays to be able to handle at least "x" amount of evidence.
+    private void ensureLikelihoodsMatrixEvidenceCapacity(final int sampleIndex, final int x,
+                                                         final double[][] sampleValues, final int numberOfAlleles) {
+        final int currentCapacity = likelihoodsMatrixEvidenceCapacityBySampleIndex[sampleIndex];
+        if (currentCapacity < x) {
+            final int newCapacity = Math.max(currentCapacity, x) << 1; // we double it to avoid repetitive 1-element extensions resizing.
+            for (int a  = 0; a < numberOfAlleles; a++) {
+                sampleValues[a] = Arrays.copyOf(sampleValues[a], newCapacity);
+                // We don't need the following fill for this to work as intended but
+                // with NaN we ensure some clear error output in case there is a bug that
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoods.java:815]: clear and early.

File: src/test/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoodsUnitTest.java [hunk 13]
```diff
@@ -26,6 +26,23 @@
     private static final int ODD_READ_START = 101;
     private static final int EVEN_READ_START = 1;
 
+
+    private static final SAMFileHeader SAM_HEADER = ArtificialReadUtils.createArtificialSamHeader(10, 0, 1000);
+    private static final GenomeLocParser locParser = new GenomeLocParser(SAM_HEADER.getSequenceDictionary());
+
```
Comment (vruano) [inline src/test/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoodsUnitTest.java:32]: extra blank line

File: src/test/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoodsUnitTest.java [hunk 14]
```diff
@@ -26,6 +26,23 @@
     private static final int ODD_READ_START = 101;
     private static final int EVEN_READ_START = 1;
 
+
```
Comment (vruano) [inline src/test/java/org/broadinstitute/hellbender/utils/genotyper/AlleleLikelihoodsUnitTest.java:29]: extra blank line.

File: src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.testGVCFMode.gatk4.alleleSpecific.g.vcf [hunk 15]
```diff
@@ -240,7 +240,7 @@
 20	10005724	.	A	<NON_REF>	.	.	END=10006290	GT:DP:GQ:MIN_DP:PL	0/0:82:99:53:0,120,1800
 20	10006291	.	G	A,<NON_REF>	999.64	.	AS_RAW_BaseQRankSum=|0.3,1|NaN;AS_RAW_MQ=188569.00|106610.00|0.00;AS_RAW_MQRankSum=|-1.1,1|NaN;AS_RAW_ReadPosRankSum=|-1.5,1|NaN;AS_SB_TABLE=25,28|16,15|0,0;BaseQRankSum=0.311;DP=85;ExcessHet=3.0103;MLEAC=1,0;MLEAF=0.500,0.00;MQRankSum=-1.081;RAW_MQandDP=298779,85;ReadPosRankSum=-1.447	GT:AD:DP:GQ:PL:SB	0/1:53,31,0:84:99:1007,0,1905,1167,1998,3165:25,28,16,15
 20	10006292	.	A	<NON_REF>	.	.	END=10006403	GT:DP:GQ:MIN_DP:PL	0/0:81:99:76:0,120,1800
-20	10006404	.	A	C,<NON_REF>	2943.06	.	AS_RAW_BaseQRankSum=||;AS_RAW_MQ=0.00|268082.00|1369.00;AS_RAW_MQRankSum=||;AS_RAW_ReadPosRankSum=||;AS_SB_TABLE=0,0|27,49|1,0;DP=83;ExcessHet=3.0103;MLEAC=2,0;MLEAF=1.00,0.00;RAW_MQandDP=291051,83	GT:AD:DP:GQ:PL:SB	1/1:0,76,1:77:99:2957,228,0,2957,228,2957:0,0,28,49
```
Comment (vruano) [inline src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.testGVCFMode.gatk4.alleleSpecific.g.vcf:243]: This is an interesting lateral effect fix..... there were some reads that supported the NON_REF that should not be there as per policy .... that was result of this very same bug as filtered reads were supporting the NON_REF just at random. 

So what we should see in this this test vcf updates is some (usually) small changes in AD and annotations that depend on read likelihoods across all alleles in general and specifically for NON_REF where spurious "supporting reads" will disappear (which is the right thing).

**Comment by vruano** [comment]
Squashing and rebasing, will merge as soon as test pass.

**Pull request merged**